### PR TITLE
adds updates for master deployments

### PR DESCRIPTION
**Pull request by xh3b4sd**
Towards https://github.com/giantswarm/giantswarm/issues/3272.

### CODE REVIEW

File: service/controller/v2/key/key.go [hunk 1]
```diff
@@ -158,19 +158,19 @@ func WorkerSubnetName(customObject providerv1alpha1.AzureConfig) string {
 	return fmt.Sprintf("%s-%s-%s", ClusterID(customObject), virtualNetworkSuffix, workerSubnetSuffix)
 }
 
-func MasterVMSSName(customObject providerv1alpha1.AzureConfig) string {
-	return fmt.Sprintf("%s-master", ClusterID(customObject))
-}
-
-func WorkerVMSSName(customObject providerv1alpha1.AzureConfig) string {
-	return fmt.Sprintf("%s-worker", ClusterID(customObject))
+func MasterInstanceName(customObject providerv1alpha1.AzureConfig, instanceID string) string {
```
Comment (MarcelMue) [inline service/controller/v2/key/key.go]: Would be cool to align the godoc in this file. Half of the functions have it, the other half doesn't.
No blocker though <3

File: service/controller/v2/key/key.go [hunk 2]
```diff
@@ -158,19 +158,19 @@ func WorkerSubnetName(customObject providerv1alpha1.AzureConfig) string {
...
+	return fmt.Sprintf("%s-master-%06s", ClusterID(customObject), instanceID)
 }
 
 // MasterNICName returns name of the master NIC.
 func MasterNICName(customObject providerv1alpha1.AzureConfig) string {
 	return fmt.Sprintf("%s-Master-1-NIC", ClusterID(customObject))
```
Comment (tuommaki) [inline service/controller/v2/key/key.go]: I know this line is not part of this PR, but is there a reason why master here is written with uppercase M when others are lowercase? If it can also be lowercase, maybe it could be aligned here?

Comment (xh3b4sd) [reply service/controller/v2/key/key.go]: No idea.

**Comment by xh3b4sd** [comment]
@r7vme these changes cause issues I am not sure how to debug. I have a cluster `53dsb` on `godsmack`. The issue is reproducible in a pretty straight forward way. Just create a cluster and the master gets fucked. There seems to be a networking issue. I think this is related to the issues you already discovered today. Do you want to have a look? I am not sure where to go from here. 

```
$ xh3b4sd@master0 ~ $ ssh 10.1.0.6
The authenticity of host '10.1.0.6 (10.1.0.6)' can't be established.
ECDSA key fingerprint is SHA256:s2nSyGcwBnOMR8FqeGdMOPb2gJ9JA4GmN71CpEXXUI0.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added '10.1.0.6' (ECDSA) to the list of known hosts.
Container Linux by CoreOS stable (1745.5.0)
Update Strategy: No Reboots
Failed Units: 2
  k8s-addons.service
  k8s-ingress-loadbalancer.service
```

```
$ xh3b4sd@53dsb-master-000001 ~ $ sudo journalctl -fu k8s-addons.service
-- Logs begin at Mon 2018-06-11 15:31:46 UTC. --
Jun 11 15:36:16 53dsb-master-000001 systemd[1]: k8s-addons.service: Failed to load environment files: No such file or directory
Jun 11 15:36:16 53dsb-master-000001 systemd[1]: k8s-addons.service: Failed to run 'start' task: No such file or directory
Jun 11 15:36:16 53dsb-master-000001 systemd[1]: k8s-addons.service: Failed with result 'resources'.
Jun 11 15:36:16 53dsb-master-000001 systemd[1]: Failed to start Kubernetes Addons.
-- Reboot --
Jun 11 15:48:25 53dsb-master-000001 systemd[1]: k8s-addons.service: Failed to load environment files: No such file or directory
Jun 11 15:48:25 53dsb-master-000001 systemd[1]: k8s-addons.service: Failed to run 'start' task: No such file or directory
Jun 11 15:48:25 53dsb-master-000001 systemd[1]: k8s-addons.service: Failed with result 'resources'.
Jun 11 15:48:25 53dsb-master-000001 systemd[1]: Failed to start Kubernetes Addons.
...
```

```
$ xh3b4sd@53dsb-master-000001 ~ $ sudo journalctl -fu k8s-ingress-loadbalancer.service
-- Logs begin at Mon 2018-06-11 15:31:46 UTC. --
Jun 11 15:36:32 53dsb-master-000001 sh[1231]: Digest: sha256:995bd3fee6899569d5d9ab77948f25d6bc2e9a95efa988de37c6b8c3095ac819
Jun 11 15:36:32 53dsb-master-000001 sh[1231]: Status: Downloaded newer image for quay.io/giantswarm/docker-kubectl:8cabd75bacbcdad7ac5d85efc3ca90c2fabf023b
Jun 11 15:37:03 53dsb-master-000001 sh[1231]: W0611 15:37:03.756008       1 factory_object_mapping.go:423] Failed to download OpenAPI (Get https://api.53dsb.k8s.godsmack.westeurope.azure.gigantic.io/swagger-2.0.0.pb-v1: dial tcp 40.118.42.141:443: i/o timeout), falling back to swagger
Jun 11 15:38:03 53dsb-master-000001 sh[1231]: Unable to connect to the server: dial tcp 40.118.42.141:443: i/o timeout
-- Reboot --
Jun 11 15:48:25 53dsb-master-000001 systemd[1]: Starting Script to create Kubernetes load balancer...
Jun 11 15:51:36 53dsb-master-000001 sh[719]: W0611 15:51:26.803879       1 factory_object_mapping.go:423] Failed to download OpenAPI (Get https://api.53dsb.k8s.godsmack.westeurope.azure.gigantic.io/swagger-2.0.0.pb-v1: dial tcp 40.118.42.141:443: i/o timeout), falling back to swagger
Jun 11 15:52:50 53dsb-master-000001 sh[719]: Unable to connect to the server: dial tcp 40.118.42.141:443: i/o timeout
Jun 11 15:53:45 53dsb-master-000001 systemd[1]: k8s-ingress-loadbalancer.service: Main process exited, code=exited, status=1/FAILURE
Jun 11 15:53:45 53dsb-master-000001 systemd[1]: k8s-ingress-loadbalancer.service: Failed with result 'exit-code'.
Jun 11 15:53:45 53dsb-master-000001 systemd[1]: Failed to start Script to create Kubernetes load balancer.
...
```

```
$ xh3b4sd@53dsb-master-000001 ~ $ sudo journalctl -fu etcd3
-- Logs begin at Mon 2018-06-11 15:31:46 UTC. --
Jun 11 15:59:00 53dsb-master-000001 docker[1428]: 2018-06-11 15:59:00.323492 W | etcdserver: read-only range request "key:\"/giantswarm.io/secrets/kube-system/cronjob-controller-token-pp449\" " took too long (209.240411ms) to execute
Jun 11 15:59:02 53dsb-master-000001 docker[1428]: 2018-06-11 15:59:02.483312 I | embed: rejected connection from "168.63.129.16:52111" (error "EOF", ServerName "")
Jun 11 15:59:07 53dsb-master-000001 docker[1428]: 2018-06-11 15:59:07.485100 I | embed: rejected connection from "168.63.129.16:52240" (error "EOF", ServerName "")
Jun 11 15:59:12 53dsb-master-000001 docker[1428]: 2018-06-11 15:59:12.485678 I | embed: rejected connection from "168.63.129.16:52285" (error "EOF", ServerName "")
Jun 11 15:59:17 53dsb-master-000001 docker[1428]: 2018-06-11 15:59:17.486005 I | embed: rejected connection from "168.63.129.16:52350" (error "EOF", ServerName "")
Jun 11 15:59:20 53dsb-master-000001 docker[1428]: 2018-06-11 15:59:20.350172 W | etcdserver: read-only range request "key:\"/giantswarm.io/networkpolicies\" range_end:\"/giantswarm.io/networkpoliciet\" count_only:true " took too long (154.629925ms) to execute
Jun 11 15:59:22 53dsb-master-000001 docker[1428]: 2018-06-11 15:59:22.487063 I | embed: rejected connection from "168.63.129.16:52419" (error "EOF", ServerName "")
Jun 11 15:59:27 53dsb-master-000001 docker[1428]: 2018-06-11 15:59:27.488496 I | embed: rejected connection from "168.63.129.16:52488" (error "EOF", ServerName "")
Jun 11 15:59:32 53dsb-master-000001 docker[1428]: 2018-06-11 15:59:32.489254 I | embed: rejected connection from "168.63.129.16:52542" (error "EOF", ServerName "")
Jun 11 15:59:32 53dsb-master-000001 docker[1428]: 2018-06-11 15:59:32.897087 W | etcdserver: read-only range request "key:\"/giantswarm.io/deployments\" range_end:\"/giantswarm.io/deploymentt\" count_only:true " took too long (126.37189ms) to execute
...
```

```
$ sudo journalctl -fu k8s-kubelet.service
-- Logs begin at Mon 2018-06-11 15:31:46 UTC. --
Jun 11 15:59:03 53dsb-master-000001 sh[1513]: E0611 15:59:03.021803    1538 kubelet.go:2125] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Jun 11 15:59:08 53dsb-master-000001 sh[1513]: W0611 15:59:08.023054    1538 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Jun 11 15:59:08 53dsb-master-000001 sh[1513]: E0611 15:59:08.024112    1538 kubelet.go:2125] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Jun 11 15:59:09 53dsb-master-000001 sh[1513]: E0611 15:59:09.497762    1538 kubelet.go:1617] Failed creating a mirror pod for "k8s-controller-manager-53dsb-master-000001_kube-system(1a85b362c9cd78ce249093845cee57af)": pods "k8s-controller-manager-53dsb-master-000001" is forbidden: no providers available to validate pod request
Jun 11 15:59:10 53dsb-master-000001 sh[1513]: I0611 15:59:10.329408    1538 kubelet_node_status.go:434] Using node IP: "10.1.0.6"
Jun 11 15:59:10 53dsb-master-000001 sh[1513]: I0611 15:59:10.353179    1538 azure_backoff.go:129] backoff: success
Jun 11 15:59:13 53dsb-master-000001 sh[1513]: W0611 15:59:13.025267    1538 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Jun 11 15:59:13 53dsb-master-000001 sh[1513]: E0611 15:59:13.025533    1538 kubelet.go:2125] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Jun 11 15:59:18 53dsb-master-000001 sh[1513]: W0611 15:59:18.026518    1538 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Jun 11 15:59:18 53dsb-master-000001 sh[1513]: E0611 15:59:18.026646    1538 kubelet.go:2125] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Jun 11 15:59:20 53dsb-master-000001 sh[1513]: I0611 15:59:20.364515    1538 kubelet_node_status.go:434] Using node IP: "10.1.0.6"
Jun 11 15:59:20 53dsb-master-000001 sh[1513]: I0611 15:59:20.390936    1538 azure_backoff.go:129] backoff: success
Jun 11 15:59:20 53dsb-master-000001 sh[1513]: E0611 15:59:20.495316    1538 kubelet.go:1617] Failed creating a mirror pod for "k8s-scheduler-53dsb-master-000001_kube-system(c92d14d7fe5feeb01297612addab55c3)": pods "k8s-scheduler-53dsb-master-000001" is forbidden: no providers available to validate pod request
...
```

**Comment by xh3b4sd** [comment]
@calvix you may can help here?

File: service/controller/v2/key/key.go (hunk 2 above)
Comment (calvix) [reply service/controller/v2/key/key.go]: I think we use lowercase for DNS/hostname but names in azure components  are usually uppercase ( as its usual in ms environment)

**Comment by calvix** [comment]
Seems like an issue with `k8s-kubelet` not having any requirements for start.
This cause `k8s-addons.service` and `k8s-ingress-loadbalancer.service`to start way too early and fail as network is not ready yet

As k8s-kubelet has restart policy, it will heal but the other units are not rerun after that.

**Comment by xh3b4sd** [comment]
That sounds like a general problem we have in k8scc. Do you know why we get this problem here now? What do you think would be the best way to approach this problem now?

**Comment by calvix** [comment]
Yes, i created this PR https://github.com/giantswarm/k8scloudconfig/pull/388

this make sure that k8s-kubelet is started only after network is avaiable and when all  files are ready.
This should make the rest of chain (`k8s-addons.service` and `k8s-ingress-loadbalancer.service`) start when  resources for them are also ready.

**Comment by xh3b4sd** [comment]
What about the root cause here? Do you have a hunch why this issue popped up now at all?

**Comment by r7vme** [comment]
Root cause can be that we masked `systemd-wait-network-online` time ago because of ipv6 related issues.

`k8s-setup-network-env.service` fails, because network is not ready. Proper solution would be to enable `systemd-wait-network-online` and add dependency `network-online.target` to `k8s-setup-network-env.service`. But we can not enable `systemd-wait-network-online`, because it's flappy.

If still does not work, i can help.

**Comment by calvix** [comment]
I dont knwo why its popped now,my guess would we somehow speed up the start of node (or slow down the speed how network is provisioned)

**Comment by r7vme** [comment]
This can also be related to https://github.com/giantswarm/azure-operator/issues/232

**Comment by xh3b4sd** [comment]
So the guest clusters come up but two things I noticed. First, the master seems to be ahead a bit in terms of kernel version and container runtime. Why is that now? 

```
$ kubectl get node -o wide
NAME                  STATUS    ROLES     AGE       VERSION   EXTERNAL-IP   OS-IMAGE                                        KERNEL-VERSION      CONTAINER-RUNTIME
zm7fy-master-000001   Ready     master    4m        v1.10.4   <none>        Container Linux by CoreOS 1745.6.0 (Rhyolite)   4.14.48-coreos-r1   docker://18.3.1
zm7fy-worker-000000   Ready     <none>    4m        v1.10.4   <none>        Container Linux by CoreOS 1688.5.3 (Rhyolite)   4.14.32-coreos      docker://17.12.1-ce
zm7fy-worker-000003   Ready     <none>    4m        v1.10.4   <none>        Container Linux by CoreOS 1688.5.3 (Rhyolite)   4.14.32-coreos      docker://17.12.1-ce
zm7fy-worker-000004   Ready     <none>    4m        v1.10.4   <none>        Container Linux by CoreOS 1688.5.3 (Rhyolite)   4.14.32-coreos      docker://17.12.1-ce
```

Second, in the portal I saw an issue I explained in https://github.com/giantswarm/azure-operator/issues/233#issuecomment-396653225. Not sure where it belongs to.

**Comment by calvix** [comment]
I have no idea about this https://github.com/giantswarm/azure-operator/issues/233#issuecomment-396653225

**Comment by xh3b4sd** [comment]
I think this is related to something Roman worked on. We recently added this native LB support on Azure. I am just not sure why we merged the support for this when we still have problems with this. @r7vme do you want to take a look or should I sort this here as well? I think master is "broken" because of this.

**Comment by xh3b4sd** [comment]
@calvix do you know why the master is versioned differently in the guest clusters now?

**Comment by r7vme** [comment]
> First, the master seems to be ahead a bit in terms of kernel version and container runtime. Why is that now?

because update-engine https://github.com/giantswarm/azure-operator/issues/232

**Comment by r7vme** [comment]
Second thing that i also wanted to raise, we can not preserve etcd data now, because we use VMSS and it manages disks *implicitly* (not visible to us), so we can not preserve data on the disk as it will be wiped together with vm.

Solutions here two:
1. use regular vm for master
2. use vmss, but create etcd disk separately. (afaik this now possible because kubernetes uses it to attach PV to vmss)

**Comment by xh3b4sd** [comment]
As I talked with Roman. I am going to merge this since master is in a bad shape anyway. We are going to continue fixing several issues now. I take https://github.com/giantswarm/azure-operator/issues/233 next. Will get back to you guys.
